signac/core/synced_collections/utils.py

```diff
@@ -0,0 +1,138 @@
+
+    Rather than attempting to directly optimize this behavior, this class provides
+    a workaround by which we can amortize the cost of type checks. Given a set
+    of types that must be resolved and a way to identify each of these (which
+    may be expensive), it maintains a local cache of all instances of a given
+    type that have previously been observed. This reduces the cost of type checking
+    to a simple dict lookup, except for the first time a new type is observed.
+
+    Parameters
+    ----------
+    abstract_type_identifiers : collections.abc.Mapping
+        A mapping from a string identifier for a group of types (e.g. ``MAPPING``)
+        to a callable that can be used to identify that type. Due to insertion order
+        guarantees of dictionaries in Python>=3.6 (officially 3.7), it is beneficial
+        to order this dictionary with the most frequently occuring types first.
+
+    Attributes
+    ----------
+    abstract_type_identifiers : Dict[str, Callable[Any, bool]]
+        A mapping from string identifiers for an abstract type to callables that
+        accepts an object and returns True if the object is of the key type and
+        False if not.
+    type_map : Dict[Type, str]
+        A mapping from concrete types to the corresponding named abstract type
+        from :attr:`type_enum`.
+
+    """
+
+    def __init__(self, abstract_type_identifiers):
+        self.abstract_type_identifiers = abstract_type_identifiers
+        self.type_map = {}
+
+    def get_type(self, obj):
+        """Get the type string corresponding to this data type.
+
+        Parameters
+        ----------
+        obj : Any
+            Any object whose type to check
+
+        Returns
+        -------
+        str
+            The name of the type, where valid types are the keys of the dict
+            argument to the constructor. If the object's type cannot be identified,
+            will return ``None``.
+
+        """
+        dtype = type(obj)
+        enum_type = None
+        try:
+            enum_type = self.type_map[dtype]
+        except KeyError:
+            for adt, id_func in self.abstract_type_identifiers.items():
+                if id_func(obj):
+                    enum_type = self.type_map[dtype] = adt
+                    break
+            self.type_map[dtype] = enum_type
+
+        return enum_type
+
+
+def default(o: Any) -> Dict[str, Any]:  # noqa: D102
+    """Get a JSON-serializable version of compatible types.
+
+    This function is suitable for use with JSON-serialization tools as a way
+    to serialize :class:`SyncedCollection` objects and NumPy arrays.
+
+    Warnings
+    --------
+    JSON encoding of numpy arrays is not invertible; once encoded, reloading
+    the data will result in converting arrays to lists and numpy numbers into
+    ints or floats.
+
+    """
+    if NUMPY:
+        if isinstance(o, numpy.number):
+            return o.item()
```

Comment:
I see Codecov has marked a few places that are untested. I believe we have tests for the similar places where this is already done in signac, but we need to port those tests so they cover this synced collections code.

Reply:
The relevant tests are actually already in place. The reason that this isn't tested here any more is because numpy arrays are now converted further upstream. The new SyncedList type is much more sophisticated than the original one, and numpy arrays are converted to lists as soon as they are assigned.

I suppose that leaves a question of whether we need to bother including this functionality in the custom JSON encoder at all. I'd be fine with removing it, I just included it for completeness. What do you think?

Reply:
For now I left it in place in case it comes in handy later. I've added explicit tests of the utils (both the type resolver and the encoder) so I think we're set here.